### PR TITLE
multi: validate payment params at RPC layer

### PR DESCRIPTION
**Pull request by Crypt-iQ**
With this patch, we'll fail out earlier in the cycle in case of some wonky parameters, and not leave zombie payments in the router which currently are not cleaned up.

### CODE REVIEW

**Comment by Crypt-iQ** [comment]
Naming currently conflicts with another check in `routerrpc`, so I could change the validation function name.

File: lnrpc/routerrpc/router_backend.go [hunk 1]
```diff
@@ -837,6 +845,14 @@ func (r *RouterBackend) extractIntentFromSendRequest(
 		payIntent.DestFeatures = features
 	}
 
+	// Do bounds checking with the block padding so the router isn't
+	// left with a zombie payment in case the user messes up.
+	err = routing.ValidateCLTVLimit(payIntent.CltvLimit,
```
Comment (Roasbeef) [inline lnrpc/routerrpc/router_backend.go:850]: Style nit, proper folding would be: 
```go
err := routing.ValidateCLTVLimit(
    payIntent.CltvLimit, ...
)
```

File: rpcserver.go [hunk 2]
```diff
@@ -4370,6 +4370,14 @@ func (r *rpcServer) extractPaymentIntent(rpcPayReq *rpcPaymentRequest) (rpcPayme
 		payIntent.cltvDelta = uint16(r.cfg.Bitcoin.TimeLockDelta)
 	}
 
+	// Do bounds checking with the block padding so the router isn't left
+	// with a zombie payment in case the user messes up.
+	err = routing.ValidateCLTVLimit(payIntent.cltvLimit,
```
Comment (Roasbeef) [inline rpcserver.go:4375]: Same here re folding.

File: routing/payment_session.go [hunk 3]
```diff
@@ -14,6 +14,22 @@ import (
 // to prevent an HTLC being failed if some blocks are mined while it's in-flight.
 const BlockPadding uint16 = 3
 
+// ValidateCLTVLimit is a helper function that validates that the cltv limit is
+// less than the final cltv delta parameter, optionally including the
+// BlockPadding in this calculation.
+func ValidateCLTVLimit(limit uint32, delta uint16, includePad bool) error {
```
Comment (Roasbeef) [inline routing/payment_session.go:20]: Let's add some basic TDD tests for this function?

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: routing/payment_session.go [hunk 4]
```diff
@@ -14,6 +14,22 @@ import (
 // to prevent an HTLC being failed if some blocks are mined while it's in-flight.
 const BlockPadding uint16 = 3
 
+// ValidateCLTVLimit is a helper function that validates that the cltv limit is
+// less than the final cltv delta parameter, optionally including the
```
Comment (yyforyongyu) [inline routing/payment_session.go:18]: nit: I think what you mean here is that we want to validate the cltv limit is *greater* than the final cltv delta?

File: lnrpc/routerrpc/router_backend_test.go [hunk 5]
```diff
@@ -205,6 +205,7 @@ func testQueryRoutes(t *testing.T, useMissionControl bool, useMsat bool) {
 			}
 			return route.Vertex{1}, route.Vertex{2}, nil
 		},
+		MaxTotalTimelock: 1000,
```
Comment (yyforyongyu) [inline lnrpc/routerrpc/router_backend_test.go:208]: Why the change here?

**Review by yyforyongyu — CHANGES_REQUESTED** [review]

File: lnrpc/routerrpc/router_backend_test.go (hunk 5 above)
Comment (Crypt-iQ) [reply lnrpc/routerrpc/router_backend_test.go:208]: Needs to be specified so the test sets `CltvLimit`, otherwise it will default to 0

File: routing/payment_session.go (hunk 4 above)
Comment (Crypt-iQ) [reply routing/payment_session.go]: Changed comment

File: routing/payment_session.go (hunk 3 above)
Comment (Crypt-iQ) [reply routing/payment_session.go:20]: Basic tests added

File: rpcserver.go (hunk 2 above)
Comment (Crypt-iQ) [reply rpcserver.go]: Function folded

File: lnrpc/routerrpc/router_backend.go (hunk 1 above)
Comment (Crypt-iQ) [reply lnrpc/routerrpc/router_backend.go]: Function folded

File: routing/payment_session_test.go [hunk 6]
```diff
@@ -8,6 +8,54 @@ import (
 	"github.com/lightningnetwork/lnd/routing/route"
 )
 
+func TestValidateCLTVLimit(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		cltvLimit      uint32
```
Comment (yyforyongyu) [inline routing/payment_session_test.go:15]: Could add a test name here to give a rough idea of what case is under test.

File: routing/payment_session_test.go [hunk 7]
```diff
@@ -8,6 +8,54 @@ import (
...
+			expectError:    false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		err := ValidateCLTVLimit(
+			testCase.cltvLimit, testCase.finalCltvDelta,
+			testCase.includePadding,
+		)
+		if testCase.expectError && err == nil {
```
Comment (yyforyongyu) [inline routing/payment_session_test.go:51]: I think we want to use `require` wherever we can in tests now, can be changed to sth like,
```go
if expected error:
    require.Error(t, err, "...")
    continue

require.NoError(t, err, "...")
```

File: routing/payment_session_test.go [hunk 8]
```diff
@@ -8,6 +8,54 @@ import (
...
+		},
+		{
+			cltvLimit:      uint32(101),
+			finalCltvDelta: uint16(100),
+			includePadding: false,
+			expectError:    false,
+		},
+	}
+
+	for _, testCase := range testCases {
```
Comment (yyforyongyu) [inline routing/payment_session_test.go:46]: if it's meant to use [table drive test](https://github.com/golang/go/wiki/TableDrivenTests), you'll want sth like,
```
tc := testCase
t.Run(tt.name, func(t *testing.T) {
    ...
}
```

File: lnrpc/routerrpc/router_backend_test.go (hunk 5 above)
Comment (yyforyongyu) [reply lnrpc/routerrpc/router_backend_test.go:208]: I see, could we also test that the error is returned in this test?

**Review by yyforyongyu — CHANGES_REQUESTED** [review]

File: routing/payment_session_test.go (hunk 8 above)
Comment (Crypt-iQ) [reply routing/payment_session_test.go:52]: thanks, forgot to do it like this

File: routing/payment_session_test.go (hunk 7 above)
Comment (Crypt-iQ) [reply routing/payment_session_test.go]: using require

File: routing/payment_session_test.go (hunk 6 above)
Comment (Crypt-iQ) [reply routing/payment_session_test.go:17]: added

File: lnrpc/routerrpc/router_backend_test.go (hunk 5 above)
Comment (Crypt-iQ) [reply lnrpc/routerrpc/router_backend_test.go]: Optionally setting it now to see if error is returned

**Review by Roasbeef — APPROVED** [review]
LGTM 🐴
